### PR TITLE
Change how get threads and get replies work

### PR DESCRIPTION
**Pull request by kimyen**
adding param to filter the kind of threads/ replies to return

### CODE REVIEW

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java [hunk 1]
```diff
@@ -7332,32 +7331,11 @@ public DiscussionThreadBundle getThread(String threadId)
...
-	}
-
-	private PaginatedResults<DiscussionThreadBundle> doGetThreads(
-			String forumId, Long limit, Long offset,
-			DiscussionThreadOrder order, Boolean ascending, String threadType)
-			throws SynapseException, SynapseClientException {
+			Boolean ascending, DiscussionFilter filter) throws SynapseException {
 		ValidateArgument.required(forumId, "forumId");
 		ValidateArgument.required(limit, "limit");
 		ValidateArgument.required(offset, "offset");
```
Comment (kimyen) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java]: validate filter

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/discussion/DBODiscussionReplyDAOImplTest.java [hunk 2]
```diff
@@ -359,4 +390,85 @@ public void testGetThreadAuthorStats() throws InterruptedException {
...
+				+ " DISCUSSION_REPLY.ETAG AS ETAG, DISCUSSION_REPLY.IS_EDITED AS IS_EDITED,"
+				+ " DISCUSSION_REPLY.IS_DELETED AS IS_DELETED"
+				+ " FROM DISCUSSION_REPLY, DISCUSSION_THREAD, FORUM"
+				+ " WHERE THREAD_ID = DISCUSSION_THREAD.ID"
+				+ " AND FORUM_ID = FORUM.ID"
+				+ " AND THREAD_ID = ?"
+				+ " ORDER BY CREATED_ON DESC"
+				+ " LIMIT 10 OFFSET 3",
+				DBODiscussionReplyDAOImpl.buildGetRepliesQuery(10L, 3L, DiscussionReplyOrder.CREATED_ON, false, DiscussionFilter.NO_FILTER));
+	}
```
Comment (kimyen) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/discussion/DBODiscussionReplyDAOImplTest.java]: add test for addCondition()

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/discussion/DBODiscussionReplyDAOImplTest.java [hunk 3]
```diff
@@ -137,53 +138,63 @@ public void testCreate() {
...
 	public void testGetRepliesForThreadWithNullOrderNotNullAscending() {
-		replyDao.getRepliesForThread(threadIdLong, 1L, 0l, null, true);
+		replyDao.getRepliesForThread(threadIdLong, 1L, 0l, null, true, DiscussionFilter.NO_FILTER);
 	}
 
 	@Test (expected = IllegalArgumentException.class)
 	public void testGetRepliesForThreadWithNotNullOrderNullAscending() {
-		replyDao.getRepliesForThread(threadIdLong, 1L, 0l, DiscussionReplyOrder.CREATED_ON, null);
+		replyDao.getRepliesForThread(threadIdLong, 1L, 0l, DiscussionReplyOrder.CREATED_ON, null, DiscussionFilter.NO_FILTER);
 	}
```
Comment (kimyen) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/discussion/DBODiscussionReplyDAOImplTest.java]: add unauthorized test

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/discussion/DBODiscussionThreadDAOImplTest.java [hunk 4]
```diff
@@ -202,7 +203,7 @@ public void testUpdateMessageUrlWithInvalidArgument(){
 	public void testGetThreadsWithZeroExistingThreads() {
 		assertEquals("empty threads",
 				new ArrayList<DiscussionThreadBundle>(),
-				threadDao.getThreads(forumIdLong, MAX_LIMIT, 0L, null, null).getResults());
+				threadDao.getThreads(forumIdLong, MAX_LIMIT, 0L, null, null, DiscussionFilter.NO_FILTER).getResults());
 	}
```
Comment (kimyen) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/discussion/DBODiscussionThreadDAOImplTest.java]: add unauthorized test

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/discussion/DiscussionThreadManagerImpl.java [hunk 5]
```diff
@@ -78,7 +79,7 @@ public DiscussionThreadBundle getThread(UserInfo userInfo, String threadId) {
 		AuthorizationManagerUtil.checkAuthorizationAndThrowException(
 				authorizationManager.canAccess(userInfo, thread.getProjectId(), ObjectType.ENTITY, ACCESS_TYPE.READ));
 		threadDao.updateThreadView(threadIdLong, userInfo.getId());
-		return updateNumberOfReplies(thread);
+		return updateNumberOfReplies(thread, DiscussionFilter.NOT_DELETED_ONLY);
```
Comment (kimyen) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/discussion/DiscussionThreadManagerImpl.java]: add test for this

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/discussion/DiscussionThreadManagerImpl.java [hunk 6]
```diff
@@ -78,7 +79,7 @@ public DiscussionThreadBundle getThread(UserInfo userInfo, String threadId) {
 		AuthorizationManagerUtil.checkAuthorizationAndThrowException(
```
Comment (kimyen) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/discussion/DiscussionThreadManagerImpl.java]: only return not deleted threads

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/discussion/DiscussionThreadManagerImplTest.java [hunk 7]
```diff
@@ -235,53 +245,19 @@ public void testDeleteAuthorized() {
...
-		Mockito.when(mockThreadDao.getDeletedThreads(Mockito.anyLong(), Mockito.anyLong(), Mockito.anyLong(), (DiscussionThreadOrder) Mockito.any(), Mockito.anyBoolean()))
-				.thenReturn(threads);
-		Mockito.when(mockAuthorizationManager.canAccess(userInfo, projectId, ObjectType.ENTITY, ACCESS_TYPE.READ))
+		Mockito.when(mockAuthorizationManager.canAccess(userInfo, projectId, ObjectType.ENTITY, ACCESS_TYPE.MODERATE))
 				.thenReturn(AuthorizationManagerUtil.AUTHORIZED);
-		assertEquals(threads, threadManager.getDeletedThreadsForForum(userInfo, forumId.toString(), 2L, 0L, DiscussionThreadOrder.LAST_ACTIVITY, true));
-		Mockito.verify(mockReplyDao).getReplyCount(threadId);
+		assertEquals(threads, threadManager.getThreadsForForum(userInfo, forumId.toString(), 2L, 0L, DiscussionThreadOrder.LAST_ACTIVITY, true, DiscussionFilter.NO_FILTER));
+		Mockito.verify(mockReplyDao).getReplyCount(threadId, DiscussionFilter.NO_FILTER);
 	}
```
Comment (kimyen) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/discussion/DiscussionThreadManagerImplTest.java]: add test case when user is not a moderator

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/discussion/DiscussionFilter.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+    "description":"The filter that helps determine what threads/ replies to return.",
+    "name": "DiscussionFilter",
+    "type":"string",
+    "enum":[
+        "NO_FILTER",
+        "DELETED_ONLY",
+        "NOT_DELETED_ONLY"
+    ]
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/discussion/DiscussionFilter.json:9]: maybe NOT_DELETED_ONLY to EXCLUDE_DELETED

**Pull request merged**